torch/csrc/autograd/python_function.cpp

```diff
@@ -844,7 +912,12 @@ static PyObject *unpack_saved_variables(
   THPObjectPtr saved(PyTuple_New(num_saved));
   if (!saved)
     return nullptr;
-  auto saved_for = THPFunction_asFunction(self);
+  auto saved_for = self->cdata.lock();
+  // This is really a true assert, because we've already tested for the
+  // self->has_freed_buffers case at the beginning of this function:
+  // buffers are freed when PyFunction dies; if the buffers are not freed,
+  // PyFunction must be live.
```

Comment:
Hmmm are you sure? I think those are orthogonal conditions. `has_freed_buffers` is only triggered by running backward, but you can have `cdata` get freed with no backward runs involved.

Reply:
IIUC, it's a superset condition. If PyFunction is destructed, `has_freed_buffers` is guaranteed to be true, but not vice versa. That's good enough here.